### PR TITLE
Feature/000 add interfaces

### PR DESCRIPTION
**Pull request by mgazza**
gocui is lacking abstraction making it very hard to test code that consumes it directly. 
This change adds two interfaces over gocui to enable testing

### CODE REVIEW

**Comment by mgazza** [comment]
Apologies for the naming of the interfaces. I wanted to make it easy to identify and therefore refactor in consuming code. If you can suggest better alternatives please do :)

File: gui.go [hunk 1]
```diff
@@ -136,25 +194,24 @@ func (g *Gui) SetView(name string, x0, y0, x1, y1 int) (*View, error) {
...
+		v.SetBounds(x0, x1, y0, y1)
+		v.Invalidate()
 		return v, nil
 	}
 
 	v := newView(name, x0, y0, x1, y1, g.outputMode)
-	v.BgColor, v.FgColor = g.BgColor, g.FgColor
-	v.SelBgColor, v.SelFgColor = g.SelBgColor, g.SelFgColor
+	v.SetBgFgColor(g.BgColor, g.FgColor)
+	//v.BgColor, v.FgColor = g.BgColor, g.FgColor
```
Comment (damastes) [inline gui.go]: remove commented old code

File: gui.go [hunk 2]
```diff
@@ -239,14 +299,14 @@ func (g *Gui) SetCurrentView(name string) (*View, error) {
...
-func (g *Gui) CurrentView() *View {
+func (g *Gui) CurrentView() Viewer {
 	return g.currentView
 }
 
 // SetKeybinding creates a new keybinding. If viewname equals to ""
 // (empty string) then the keybinding will apply to all views. key must
 // be a rune or a Key.
-func (g *Gui) SetKeybinding(viewname string, key interface{}, mod Modifier, handler func(*Gui, *View) error) error {
+func (g *Gui) SetKeybinding(viewname string, key interface{}, mod Modifier, handler func(Guier, Viewer) error) error {
```
Comment (damastes) [inline gui.go:307]: Guier is a poop name

File: gui.go [hunk 3]
```diff
@@ -164,27 +221,27 @@ func (g *Gui) SetViewOnTop(name string) (*View, error) {
 }
 
 // SetViewOnBottom sets the given view on bottom of the existing ones.
-func (g *Gui) SetViewOnBottom(name string) (*View, error) {
+func (g *Gui) SetViewOnBottom(name string) (Viewer, error) {
 	for i, v := range g.views {
-		if v.name == name {
+		if v.Name() == name {
```
Comment (damastes) [inline gui.go:224]: Maps exist

**Review by damastes — CHANGES_REQUESTED** [review]

File: gui.go (hunk 3 above)
Comment (mgazza) [reply gui.go:224]: yea was tempted to change this. it appears whoever wrote this didn't

File: gui.go (hunk 2 above)
Comment (mgazza) [reply gui.go:307]: indeed; it is in keeping with go interface naming, i.e. reader, writer. I'd like to keep it upstream mergeable so I'm open to suggestions for the interface names

**Review by damastes — APPROVED** [review]

**Review by skldt — APPROVED** [review]
Nice :) About the naming though, I think it's okay to move away from Go's convention of naming interfaces here.

Maybe `Viewer -> Widget`? And `Guier -> Application`?

**Comment by mgazza** [comment]
Going to keep the terrible interface names at present and just get this in. IMO view and gui should be renamed as well as the interfaces as they are also non-descriptive of what they do.

**Pull request merged**